File: 20231101_Wednesday/20231101.go
// Leetcode: 501. Find Mode in Binary Search Tree.
// https://leetcode.com/problems/find-mode-in-binary-search-tree/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 12 ms, faster than 37.88% of Go online submissions for Find Mode in
// Binary Search Tree.
// Memory Usage: 6.7 MB, less than 45.45% of Go online submissions for Find Mode
// in Binary Search Tree.
// 2023.11.01 Daily Challenge.
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var c map[int]int

func findMode(root *TreeNode) []int {
	c = make(map[int]int)
	dfs(root)
	if 0 == len(c) {
		return []int{}
	}
	flag := true
	max := 0
	for _, v := range c {
		if flag {
			max = v
			flag = false
		}
		if v > max {
			max = v
		}
	}
	ans := make([]int, 0)
	for i, v := range c {
		if v == max {
			ans = append(ans, i)
		}
	}
	return ans
}

func dfs(root *TreeNode) {
	if nil == root {
		return
	}
	c[root.Val] += 1
	dfs(root.Left)
	dfs(root.Right)
}

File: 20231101_Wednesday/20231101.rb
# Leetcode: 501. Find Mode in Binary Search Tree.
# https://leetcode.com/problems/find-mode-in-binary-search-tree/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 86 ms, faster than 36.36% of Ruby online submissions for Find Mode in
# Binary Search Tree.
# Memory Usage: 213.8 MB, less than 9.09% of Ruby online submissions for Find
# Mode in Binary Search Tree.
# 2023.11.01 Daily Challenge.
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def find_mode(root)
  @c = {}
  dfs(root)
  return [] if @c.empty?
  max = @c.values.max
  @c.select{|i,x| x == max}.map(&:first)
end

def dfs(x)
  return if !x
  @c[x.val] ||= 0
  @c[x.val] += 1
  dfs(x.left)
  dfs(x.right)
end
